osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -62,12 +63,19 @@ public DifficultyAttributes Calculate([NotNull] IEnumerable<Mod> mods, Cancellat
             if (!Beatmap.HitObjects.Any())
                 return CreateDifficultyAttributes(Beatmap, playableMods, skills, clockRate);
 
-            foreach (var hitObject in getDifficultyHitObjects())
+            List<DifficultyHitObject> hitObjects = getDifficultyHitObjects().ToList();
```

Comment:
Can avoid making it a list with a previous local:
```csharp
            DifficultyHitObject last = null;

            foreach (var current in getDifficultyHitObjects())
            {
                current.PreviousBacking = new ObjectLink<DifficultyHitObject>(current, last?.PreviousBacking);

                foreach (var skill in skills)
                {
                    cancellationToken.ThrowIfCancellationRequested();
                    skill.ProcessInternal(current);
                }

                last = current;
            }
```

Same applies to the other one in `CalculateTimed`.

Reply:
This was the original setup, but I changed it to a for loop just out of elegance - can revert if there's a preference

Reply:
The bigger issue here is that each calculation method is manually setting up the `ObjectLink`s. It is `getDifficultyHitObjects()`'s responsibility as a method to set these links up to be valid, at which point all of this `.ToList()` badness goes away.